PkgConfigDependency: Sort -L flags according to PKG_CONFIG_PATH

lantw44 opened the pull request:
When there is more than one path in PKG_CONFIG_PATH. It is almost always
preferred to find things in the order specified by PKG_CONFIG_PATH
instead of assuming pkg-config returns flags in a meaningful order.

For example:

/usr/local/lib/libgtk-3.so.0
/usr/local/lib/pkgconfig/gtk+-3.0.pc
/usr/local/lib/libcanberra-gtk3.so
/usr/local/lib/pkgconfig/libcanberra-gtk3.pc
/home/mesonuser/.local/lib/libgtk-3.so.0
/home/mesonuser/.local/lib/pkgconfig/gtk+-3.0.pc

PKG_CONFIG_PATH="/home/mesonuser/.local/lib/pkgconfig:/usr/local/lib/pkgconfig"

libcanberra-gtk3 is a library which depends on gtk+-3.0. The dependency
is mentioned in the .pc file with 'Requires', so flags from gtk+-3.0 are
used in both dynamic and static linking.

Assume the user wants to compile an application which needs both
libcanberra-gtk3 and gtk+-3.0. The application depends on features added
in the latest version of gtk+-3.0, which can be found in the home
directory of the user but not in /usr/local. When meson asks pkg-config
for linker flags of libcanberra-gtk3, pkg-config picks
/usr/local/lib/pkgconfig/libcanberra-gtk3.pc and
/home/mesonuser/.local/lib/pkgconfig/gtk+-3.0.pc. Since these two
libraries come from different prefixes, there will be two -L arguments
in the output of pkg-config. If -L/usr/local/lib is put before
-L/home/mesonuser/.local/lib, meson will find both libraries in
/usr/local/lib instead of picking libgtk-3.so.0 from the home directory.

This can result in linking failure such as undefined references error
when meson decides to put linker arguments of libcanberra-gtk3 before
linker arguments of gtk+-3.0. When both /usr/local/lib/libgtk-3.so.0 and
/home/mesonuser/.local/lib/libgtk-3.so.0 are present on the command
line, the linker chooses the first one and ignores the second one. If
the application needs new symbols that are only available in the second
one, the linker will throw an error because of missing symbols.

To resolve the issue, we should reorder -L flags according to
PKG_CONFIG_PATH ourselves before using it to find the full path of
library files. This makes sure that we always follow the preferences of
users, without depending on the unreliable part of pkg-config output.

Fixes https://github.com/mesonbuild/meson/issues/4271.

lantw44 commented:
It seems that I have to ping here to make it possible for the pull request to be reviewed.

jpakkane commented:
LGTM but I'd like to get input from someone with more pkg-config experience. @nirbheek?

Also, it would be nice to get a test for this. Building a full environment is probably not worth it but maybe `sort_libpaths` could be made a freestanding function and then it could be tested with string array arguments.

lantw44 commented:
@jpakkane I moved `sort_libpaths` out of a class and added a few unit tests for it. Is it good for reviewing again? I think it is a required patch for JHBuild to work properly, and the older version of the patch is also used in FreeBSD ports since version 0.48.

lantw44 commented:
@jpakkane @nirbheek Is there any news on this issue? Having to remember to apply patches every time Meson is updated in JHBuild is inconvenient.

lantw44 commented:
@jpakkane @nirbheek Ping! Is there anything I should do to get the pull request merged?

I am going to submit this patch for inclusion in JHBuild. It is annoying to have to remember to apply patches every time Meson is updated.

lantw44 commented:
@jpakkane @nirbheek Ping again! I reported the problem to pkgconf upstream: https://todo.sr.ht/~kaniini/pkgconf/7, but the upstream seems to be inactive. Can we consider merging this change to fix the meson 0.47 regression?

lantw44 commented:
Rebased on the master branch to resolve conflict. It is unfortunate that this pull request is open for so long and I have to keep downstream patches to get things build successfully.

lantw44 commented:
Thanks for your response. I hope we will be able to remove the downstream patch from FreeBSD ports and JHBuild!

xclaesse commented:
Looks good to me too.
